File: problems/0020/0020.cpp
#include <string>
#include <stack>

class Solution {
public:
    bool isValid(std::string s) {
        std::stack <char> st;
        for (char& i : s) {
            if (i == '(' || i == '{' || i == '[') {
                st.push(i);
            }
            else if (i == ')') {
                if (st.empty() || st.top() != '(') {
                    return false;
                }
                st.pop();
            }
            else if (i == '}') {
                if (st.empty() || st.top() != '{') {
                    return false;
                }
                st.pop();
            }
            else {
                if (st.empty() || st.top() != '[') {
                    return false;
                }
                st.pop();
            }
        }
        return st.empty();
    }
};

File: problems/0020/0020.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack: list[str] = []
        for c in s:
            if c == '(' or c == '{' or c == '[':
                stack.append(c)
            elif c == ')':
                if not stack or stack[-1] != '(':
                    return False
                stack.pop()
            elif c == '}':
                if not stack or stack[-1] != '{':
                    return False
                stack.pop()
            else:
                if not stack or stack[-1] != '[':
                    return False
                stack.pop()
        return not stack
